PlaceRouteHierFlow/PnR-pybind11.cpp

```diff
@@ -313,6 +314,16 @@ PYBIND11_MODULE(PnR, m) {
   //m.def("save_state", &save_state, "helper function to save_state");
   //m.def("route_single_variant", &route_single_variant, "helper function to route a single variant");
   //m.def("route_top_down", &route_top_down, "helper function to perform top-down routing");
-  m.def("toplevel", &toplevel, "helper function to perform the whole C++ flow");
 
+  m.def("toplevel", [](const std::vector<std::string>& argv) {
+    py::scoped_ostream_redirect coutstream(
+        std::cout,
+        py::module_::import("align").attr("utils").attr("logging").attr("StreamLogger")(std::string("align.pnr.toplevel"), std::string("DEBUG"))
+    );
+    py::scoped_estream_redirect cerrstream(
+        std::cerr,
+        py::module_::import("align").attr("utils").attr("logging").attr("StreamLogger")(std::string("align.pnr.toplevel"), std::string("ERROR"))
+    );
+    toplevel(argv);},
+    "helper function to perform the whole C++ flow");
```

Comment:
@stevenmburns I moved the cout/cerr to logging redirection to C++. It is cleaner because we are not modifying sys.stdout & sys.stderr. And we can control the log prefix better once we start splitting up placer & router engines (I am using align.pnr.toplevel for now)

Reply:
This "toplevel" function is going to be broken up into several C++ functions. The flow will move from the toplevel C++ code, to a Python script, calling the parsers, placer, global router, and detailed router through individual Python bindings.
We'll I guess need to do this redirection for all the individual functions, or do it once for multiple calls. Anyway, I like what you did.
A few of things:
1) it seems very slow on for example test_cmdline[adder]. But maybe this is incentive to remove silly writes to stdout.
2) it doesn't capture the output of the LP solver called by the C++ code. It is still being written to stdout.
3) Unless you merged this with master, there was some stuff being written to "stderr" that is now disappearing.

Reply:
I just sent in a commit to fix (2). Pybind only redirects cout / cerr (which are of type ostream) and not file based writes to stdout / stderr (aka printf / fprintf). We had very few of these so I just modified the printf & fprintf statements to use std::cout & std::cerr instead. Also set the lpsolve output file to /dev/null and wrote a custom log function instead.

Re (3), cerr is being redirected to the log file as well except its being marked as "ERROR" instead of "DEBUG". Can you take a look to see if something is still missing? I do see a few ERROR lines in telescopic_ota:
```
align.pnr.toplevel ERROR : PnRDB-Error: fail to read constraint file
align.pnr.toplevel ERROR : PnRDB-Error: fail to read constraint file of module
align.pnr.toplevel ERROR : telescopic_ota
```

Regarding (1) I did notice that the adder seems very slow and am trying to see how tightly correlated it is to logfile verbosity. I started by looking at the most commonly repeated messages in the log file:
```
#freq      #string
1990045 align.pnr.toplevel DEBUG : Plist problem
 233766 align.pnr.toplevel DEBUG : converter check point 2
 182125 align.pnr.toplevel DEBUG : converter check point 4
 182125 align.pnr.toplevel DEBUG : converter check point 3
 128900 align.pnr.toplevel DEBUG : converter check point 1
  53887 align.pnr.toplevel DEBUG : Enter converter
  45540 align.pnr.toplevel DEBUG : interval_number 1 1
  31978 align.pnr.toplevel DEBUG : interval_number 2 1
  31730 align.pnr.toplevel DEBUG : placedBox: LL0,0 UR0,0
  24232 align.pnr.toplevel DEBUG : 4 288 288
  19386 align.pnr.toplevel DEBUG : L shape connection 1
  19286 align.pnr.toplevel DEBUG : L shape connection 2
  18741 align.pnr.toplevel DEBUG : parallel_routing start
  18741 align.pnr.toplevel DEBUG : parallel_routing end
  18335 align.pnr.toplevel DEBUG : parallel find
  15828 align.pnr.toplevel DEBUG : 3 168 168
  15218 align.pnr.toplevel DEBUG : 2 160 160
  12436 align.pnr.toplevel DEBUG : 1 144 168
   9203 align.pnr.toplevel DEBUG : check point near node 2
   9203 align.pnr.toplevel DEBUG : check point near node 1
   9203 align.pnr.toplevel DEBUG : A start checkout point3
   8714 align.pnr.toplevel DEBUG : A start checkout point3.1
   7860 align.pnr.toplevel DEBUG : Cap Bug test 1  "PLUS"
   7860 align.pnr.toplevel DEBUG : Cap Bug test 1  "MINUS"
   5788 align.pnr.toplevel DEBUG : 0 160 160
   5549 align.pnr.toplevel DEBUG : candidate node size 1 2
   5182 align.pnr.toplevel DEBUG : originBox: LL-72,-32 UR72,32
   5167 align.pnr.toplevel DEBUG : originBox: LL-32,-72 UR32,72
   4552 align.pnr.toplevel DEBUG : @Contact  metal: M2 ; originCenter: {0,0} ; placedCenter: {0,0}
   4546 align.pnr.toplevel DEBUG : @Contact  metal: M1 ; originCenter: {0,0} ; placedCenter: {0,0}
```

I then commented out all the cout statements dealing with 'Plist problem' and 'converter'. Running profiling now.

Reply:
@Lastdayends  @854768750  we really need to take a look at the C++ cout statements. Anything that's not needed should be removed and/or a logging class should be implemented so that the verbosity can be controlled. Most of the messages look like they were added for development and then forgotten. I am pasting a list of the top 100 messages in the adder logfile below. They are all from the PnR C++ code.

```
1990045 align.pnr.toplevel DEBUG : Plist problem
 233766 align.pnr.toplevel DEBUG : converter check point 2
 182125 align.pnr.toplevel DEBUG : converter check point 4
 182125 align.pnr.toplevel DEBUG : converter check point 3
 128900 align.pnr.toplevel DEBUG : converter check point 1
  53887 align.pnr.toplevel DEBUG : Enter converter
  45540 align.pnr.toplevel DEBUG : interval_number 1 1
  31978 align.pnr.toplevel DEBUG : interval_number 2 1
  31730 align.pnr.toplevel DEBUG : placedBox: LL0,0 UR0,0
  24232 align.pnr.toplevel DEBUG : 4 288 288
  19386 align.pnr.toplevel DEBUG : L shape connection 1
  19286 align.pnr.toplevel DEBUG : L shape connection 2
  18741 align.pnr.toplevel DEBUG : parallel_routing start
  18741 align.pnr.toplevel DEBUG : parallel_routing end
  18335 align.pnr.toplevel DEBUG : parallel find
  15828 align.pnr.toplevel DEBUG : 3 168 168
  15218 align.pnr.toplevel DEBUG : 2 160 160
  12436 align.pnr.toplevel DEBUG : 1 144 168
   9203 align.pnr.toplevel DEBUG : check point near node 2
   9203 align.pnr.toplevel DEBUG : check point near node 1
   9203 align.pnr.toplevel DEBUG : A start checkout point3
   8714 align.pnr.toplevel DEBUG : A start checkout point3.1
   7860 align.pnr.toplevel DEBUG : Cap Bug test 1  "PLUS"
   7860 align.pnr.toplevel DEBUG : Cap Bug test 1  "MINUS"
   5788 align.pnr.toplevel DEBUG : 0 160 160
   5549 align.pnr.toplevel DEBUG : candidate node size 1 2
   5182 align.pnr.toplevel DEBUG : originBox: LL-72,-32 UR72,32
   5167 align.pnr.toplevel DEBUG : originBox: LL-32,-72 UR32,72
   4552 align.pnr.toplevel DEBUG : @Contact  metal: M2 ; originCenter: {0,0} ; placedCenter: {0,0}
   4546 align.pnr.toplevel DEBUG : @Contact  metal: M1 ; originCenter: {0,0} ; placedCenter: {0,0}
   3959 align.pnr.toplevel DEBUG : test 4
   3714 align.pnr.toplevel DEBUG : 5 288 288
   2866 align.pnr.toplevel DEBUG : test 2
   2197 align.pnr.toplevel DEBUG : candidate node size 1 3
   1876 align.pnr.toplevel DEBUG : west
   1876 align.pnr.toplevel DEBUG : up
   1876 align.pnr.toplevel DEBUG : south
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 9
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 8.5
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 8
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 7
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 6
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 5
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 4
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 3
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 2
   1876 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 1
   1876 align.pnr.toplevel DEBUG : north
   1876 align.pnr.toplevel DEBUG : east
   1876 align.pnr.toplevel DEBUG : down
   1874 align.pnr.toplevel DEBUG : power_grid_dsf checkpoint 8.6
   1586 align.pnr.toplevel DEBUG : test 1
   1548 align.pnr.toplevel DEBUG : Power Via Bug 1
   1527 align.pnr.toplevel DEBUG : ConvertMetal check point 4
   1527 align.pnr.toplevel DEBUG : ConvertMetal check point 3
   1527 align.pnr.toplevel DEBUG : ConvertMetal check point 2
   1527 align.pnr.toplevel DEBUG : ConvertMetal check point 1
   1288 align.pnr.toplevel DEBUG : Extract Data Step 3.6
   1288 align.pnr.toplevel DEBUG : Extract Data Step 3.5
   1288 align.pnr.toplevel DEBUG : Extract Data Step 3.4
   1288 align.pnr.toplevel DEBUG : Extract Data Step 3.3
   1288 align.pnr.toplevel DEBUG : Extract Data Step 3.2
   1288 align.pnr.toplevel DEBUG : Extract Data Step 3.1
   1071 align.pnr.toplevel DEBUG : east size 0
   1070 align.pnr.toplevel DEBUG : west size 0
   1052 align.pnr.toplevel DEBUG : candidate node size 1 1
   1046 align.pnr.toplevel DEBUG : up size -1
    984 align.pnr.toplevel DEBUG : north size 1
    982 align.pnr.toplevel DEBUG : south size 1
    894 align.pnr.toplevel DEBUG : south size 0
    892 align.pnr.toplevel DEBUG : north size 0
    830 align.pnr.toplevel DEBUG : down size -1
    806 align.pnr.toplevel DEBUG : west size 1
    805 align.pnr.toplevel DEBUG : east size 1
    502 align.pnr.toplevel DEBUG : Test 3
    502 align.pnr.toplevel DEBUG : Test 1
    501 align.pnr.toplevel DEBUG : Test 2
    493 align.pnr.toplevel DEBUG : test 3
    476 align.pnr.toplevel DEBUG : GlobalGrid-Warning: cannot find tiles to adjust vertical edge cap
    454 
    429 align.pnr.toplevel DEBUG : dest 473
    429 align.pnr.toplevel DEBUG : dest 373
    429 align.pnr.toplevel DEBUG : dest 273
    429 align.pnr.toplevel DEBUG : dest 173
    405 align.pnr.toplevel DEBUG : candidate node size 1 0
    387 align.pnr.toplevel DEBUG : temp via 5
    387 align.pnr.toplevel DEBUG : before update via
    387 align.pnr.toplevel DEBUG : after update via
    306 align.pnr.toplevel DEBUG : rect info 1 31608 -32 31752 32
    306 align.pnr.toplevel DEBUG : rect info 1 29048 -32 29192 32
    306 align.pnr.toplevel DEBUG : rect info 0 31648 -72 31712 72
    306 align.pnr.toplevel DEBUG : rect info 0 29088 -72 29152 72
    306 align.pnr.toplevel DEBUG : placedBox: LL31648,-72 UR31712,72
    306 align.pnr.toplevel DEBUG : placedBox: LL31608,-32 UR31752,32
    306 align.pnr.toplevel DEBUG : placedBox: LL29088,-72 UR29152,72
    306 align.pnr.toplevel DEBUG : placedBox: LL29048,-32 UR29192,32
    306 align.pnr.toplevel DEBUG : @Via  model:0 ; originpos: {0,0}; placedpos: {31680,0}
    306 align.pnr.toplevel DEBUG : @Via  model:0 ; originpos: {0,0}; placedpos: {29120,0}
    306 align.pnr.toplevel DEBUG : @Contact  metal: M2 ; originCenter: {0,0} ; placedCenter: {31680,0}
    306 align.pnr.toplevel DEBUG : @Contact  metal: M2 ; originCenter: {0,0} ; placedCenter: {29120,0}

```

Reply:
There is a branch #430 moving cpp print to log info.